users/services.py

```diff
@@ -0,0 +1,136 @@
+    def _send_auth_service_is_changing_notification(
+        guardian: Guardian, date_of_change_str: Optional[str] = None
+    ):
+        """Send the auth service is changing notification to a guardian.
+
+        Args:
+            guardian (Guardian): the guardian instance
+            date_of_change_str (Optional[str], optional): Date or datetime
+                in string format. Defaults to None.
+                The actual default should be given in the notification template.
+        """
+        from users.notifications import NotificationType
+
+        send_notification(
+            guardian.email,
+            NotificationType.USER_AUTH_SERVICE_IS_CHANGING,
+            context={
+                "guardian": guardian,
+                "date_of_change_str": date_of_change_str,
+                "children_event_history_markdown": (
+                    AuthServiceNotificationService.generate_children_event_history_markdown(  # noqa
+                        guardian
+                    )
+                ),
+            },
+            language=guardian.language,
+        )
+
+    @staticmethod
+    def send_user_auth_service_is_changing_notifications(
+        guardians: Optional[Union[QuerySet, list[Guardian]]] = None,
+        date_of_change_str: Optional[str] = None,
+    ):
+        """Send user authentication service is changing notifications
+        to guariands as recipients.
+
+        If the guardian list is not explicitly given
+        the queryset result of
+        `Guardian.objects.for_auth_service_is_changing_notification()`
+        will be used as a default.
+
+        Args:
+            guardians (Optional[Union[QuerySet, list[Guardian]]], optional):
+                explicit list of guardians as recipients. Defaults to None.
+            date_of_change_str (Optional[str], optional): Date or datetime
+                in string format. Defaults to None.
+                The actual default should be given in the notification template.
+        """
+        if guardians is None:
+            guardians = Guardian.objects.for_auth_service_is_changing_notification()
+
+        for guardian in guardians:
+            AuthServiceNotificationService._send_auth_service_is_changing_notification(
+                guardian, date_of_change_str
+            )
+
+    @staticmethod
+    def generate_children_event_history_markdown(guardian: Guardian):
+        """Generates a Markdown string listing a guardian's children's enrolments."""
+
+        # NOTE: This is markdown, so the line changes and white spaces are important!
+        template_string = """
+{% for child in guardian.children.all %}# {{ child.name }}
+{% for enrolment in child.enrolments.all|dictsort:"occurrence.time" %}1. **{{ enrolment.occurrence.event.name }}:** {{ enrolment.occurrence.time|date:"j.n.Y H:i" }}{% if enrolment.occurrence.event.short_description %}
+{{enrolment.occurrence.event.short_description}}{% endif %}
+{% endfor %}
+{% endfor %}
+"""  # noqa
```

Comment:
How about moving the template string outside of the class so there's no indentation so it reads clearer in relation to its surroundings' indentation level & using `.strip()` to get rid of the leading and trailing newlines unless they're necessary?
i.e.
```python
# NOTE: This is markdown, so the line changes and white spaces are important!
CHILDREN_EVENT_HISTORY_MARKDOWN_TEMPLATE = """
{% for child in guardian.children.all %}# {{ child.name }}
{% for enrolment in child.enrolments.all|dictsort:"occurrence.time" %}1. **{{ enrolment.occurrence.event.name }}:** {{ enrolment.occurrence.time|date:"j.n.Y H:i" }}{% if enrolment.occurrence.event.short_description %}
{{enrolment.occurrence.event.short_description}}{% endif %}
{% endfor %}
{% endfor %}
""".strip()  # noqa
```
How would you feel about this change?

Reply:
I also replaced the `1.` with `{{ forloop.counter }}.`, because then it looks better in the email. In the parsed markdown it does not make any change.

Reply:
MOving the strip would actually change the result a bit: The strip after the TEMPLATE would only strip the template, not the HTML that is rendered.

Reply:
The strip just strips the leading and trailing whitespace from the triple-quoted string, I mean as it starts with a newline i.e.
in Python shell:
```
>>> """
... x
... """
'\nx\n'
>>> """
... x
... """.strip()
'x'
```